Fix lyric climate

nprez83 opened the pull request:
I was having the same issues as described in #63403, specifically, the error stating that Mode 7 is not valid, only Heat, Cool, Off when trying to do anything while the thermostat is set to Auto. This error originates with the way the Lyric API handles the modes. Basically, when one queries the changeableValues dict, you get a mode=Auto, as well as a heatCoolMode, which is set to either Heat, Cool, Off. Per the documentation, heatCoolMode contains the "heat cool mode when system switch is in Auto mode". It would make sense that when changing the thermostat settings, mode=Auto should be valid, but it's not. The way the API understands that the mode should be set to Auto when changing the thermostat settings is by setting the autoChangeoverActive variable to true, not the mode itself. This require changes in the async_set_hvac_mode, async_set_temperature, and async_set_preset_mode functions. Related to this issue, I got rid of the references to hasDualSetpointStatus, as it seems that it always remains false in the API, even when the mode is set to auto, so again, the key variable for this is autoChangeoverActive.

While I was working on this I also noticed another issue. The support flag SUPPORT_TARGET_TEMPERATURE_RANGE had not been included, which did not allow for the temperature range to be available, thus invalidating the target_temperature_low and target_temperature_high functions. I added this flag and sorted out which set point (heat vs cool) should be called for each of them so things work as expected in Lovelace. I have tested all of these functionalities and they all work great on my end, so I thought I'd share.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

nprez83 committed:
Fixed the issues related to auto mode

I was having the same issues as described in #63403, specifically, the error stating that Mode 7 is not valid, only Heat, Cool, Off when trying to do anything while the thermostat is set to Auto. This error originates with the way the Lyric API handles the modes. Basically, when one queries the changeableValues dict, you get a mode=Auto, as well as a heatCoolMode, which is set to either Heat, Cool, Off. Per the documentation, heatCoolMode contains the "heat cool mode when system switch is in Auto mode". It would make sense that when changing the thermostat settings, mode=Auto should be valid, but it's not. The way the API understands that the mode should be set to Auto when changing the thermostat settings is by setting the autoChangeoverActive variable to true, not the mode itself. This require changes in the async_set_hvac_mode, async_set_temperature, and async_set_preset_mode functions. Related to this issue, I got rid of the references to hasDualSetpointStatus, as it seems that it always remains false in the API, even when the mode is set to auto, so again, the key variable for this is autoChangeoverActive.

While I was working on this I also noticed another issue. The support flag SUPPORT_TARGET_TEMPERATURE_RANGE had not been included, which did not allow for the temperature range to be available, thus invalidating the target_temperature_low and target_temperature_high functions. I added this flag and sorted out which set point (heat vs cool) should be called for each of them so things work as expected in Lovelace. I have tested all of these functionalities and they all work great on my end, so I thought I'd share.

homeassistant commented:
Hi @nprez83, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#67018).

Once you do that we will be able to review and accept this pull request.

Thanks!

nprez83 committed:
Update climate.py

nprez83 committed:
Update climate.py

Fixed two additional issues: 1) When the system is turned off from Auto, the heatCoolMode variable becomes 'Off', so when you try to restart the system back to Auto, nothing happens. 2) I now prevent the async_set_temperature function from being called with a new set point when the system is Off.
All changes tested and functional.

nprez83 commented:
Fixed two additional issues: 1) When the system is turned off from Auto, the heatCoolMode variable becomes 'Off', so when you try to restart the system back to Auto, nothing happens. 2) I now prevent the async_set_temperature function from being called with a new set point when the system is Off.
All changes tested and functional.

nprez83 committed:
Update climate.py

nprez83 committed:
Update climate.py

Return SUPPORT_PRESET_MODE flag only for LCC models (i.e. they have the "thermostatSetpointStatus" variable defined). TCC models do not support this feature

nprez83 committed:
Update climate.py

After playing with the official Honeywell API, I realized it doesn't like to received commands with missing data, i.e., it always wants to get a mode, coolSetpoint, heatSetpoint, and autoChangeoverActive variables. This was causing some random issues with changing modes, especially from coming from off, so I modified the async_set_temperature, and async_set_hvac_mode fuctions to always send all pertinent variables.

nprez83 committed:
Update climate.py

nprez83 committed:
Update climate.py

nprez83 committed:
Update climate.py

nprez83 committed:
Update climate.py

nprez83 committed:
Clean code and test everything

Alright, sorry for the multiple commits, fixing this properly took a fair bit of testing. I went ahead and cleaned up the code and made the following big picture changes: 
1) The integration now supports the Auto mode appropriately, to include the temperature range. 
2) There's a bug that actually manifests when using the native app. When the system is 'Off' and you try to turn it on to 'Auto', it will turn on briefly but will go back to 'Off' after a few seconds. When checking the web api, this appears to be related to the fact that the heatCoolMode variable seems to continue to store 'Off', even if the mode accurately displays 'Auto', and the autoChangeoverActive=True. So to overcome that inherent limitation, when the system is 'Off' and the user turns it to 'Auto', I first turn it to Heat, wait 3 seconds, and then turn it to 'Auto', which seems to work well.

nprez83 committed:
Update climate.py

nprez83 commented:
Alright, sorry for the multiple commits, fixing this properly took a fair bit of testing. I went ahead and cleaned up the code and made the following big picture changes: 
1) The integration now supports the Auto mode appropriately, to include the temperature range. 
2) There's a bug that actually manifests when using the native app. When the system is 'Off' and you try to turn it on to 'Auto', it will turn on briefly but will go back to 'Off' after a few seconds. When checking the web api, this appears to be related to the fact that the heatCoolMode variable seems to continue to store 'Off', even if the mode accurately displays 'Auto', and the autoChangeoverActive=True. So to overcome that inherent limitation, when the system is 'Off' and the user turns it to 'Auto', I first turn it to Heat, wait 3 seconds, and then turn it to 'Auto', which seems to work well.

nprez83 committed:
Fixed errors

nprez83 commented:
Sorry @timmo001, I fixed all the black, flake8, and pylint errors. I had a few flags in place for implementing the fan control, which I've done, but I don't want to incorporate that into this PR as I need to test it once the changes to aiolyric get pushed. Thanks for your help with getting these changes up and running.

nprez83 commented:
Hi @timmo001, any chance you can approve this workflow again please? The changes has been working great on my end, and it's allowed me to transfer my geofence function from the Lyric app to HA, so I'm hoping to share this with other folks as well.

timmo001 reviewed (approved):
All working well from my end. Using a T5-T6 thermostat with heating only. Still working as intended.

As for approval for the workflow, I believe a maintainer has to do this. Give them time to get through the list https://github.com/home-assistant/core/projects/4#column-5956397

nprez83 committed:
Fixed comments that were resulting in error.

nprez83 commented:
Awesome, thanks for the info and the link!

laca75tn commented:
Can you please help me?  I keep on getting the following error:
Retrying setup: 400, message='Bad Request', url=URL('https://api.honeywell.com/oauth2/token')

Thanks!

laca75tn commented:
Thanks for the response.  I deleted the integration as suggested and reinstalled it and that worked.  Thank you!
    On Monday, March 7, 2022, 01:04:36 PM EST, Numa Perez ***@***.***> wrote:  
 
 


Out of curiosity, are you running the original integration or did you manually load the code in this pull request by creating a custom component under /config/custom_components? If you've created a custom component, keep in mind you have to create a lyric folder in /config/custom_components that contains all of the original files except you replace the original climate.py with the modified one. The updates I made to climate.py do not mess with the way authorization is being handled, shouldn't give you the error you're getting.

I would suggest deleting the integration altogether and loading it up again. Before you do that, you may want to make sure the callback url you've set up in the Honeyall developer website is correct (https://developer.honeywellhome.com/user/me/apps). I'm using http://homeassistant.local:8123/auth/external/callback, but you may want to try replacing homeassistant.local with the actual local IP of your HA box, which sometimes works better. Hope it helps.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
You are receiving this because you commented.Message ID: ***@***.***>

MartinHjelmare commented on homeassistant/components/lyric/climate.py:
```diff
@@ -166,7 +170,11 @@ def __init__(
     def supported_features(self) -> int:
         """Return the list of supported features."""
-        return SUPPORT_FLAGS
+        if self.device.changeableValues.thermostatSetpointStatus:
```
Will this be constant for the lifetime of the entity? Supported features shouldn't change during entity lifetime.

nprez83 replied:
It should be constant afik. Only LCC devices will have the thermostatSetpointStatus variable defined, so only they should have the SUPPORT_PRESET_MODE flag set. If the TCC (like the Lyric) have the flag set it's no big deal, it just signals the frontend to provide the ability to change presets, which for the Lyric does nothing.

MartinHjelmare commented on homeassistant/components/lyric/climate.py:
```diff
@@ -256,48 +264,89 @@ def max_temp(self) -> float:
-                    await self._update_thermostat(
-                        self.location, device, coolSetpoint=temp
+
+        if HVAC_MODES[device.changeableValues.mode] != HVAC_MODE_OFF:
```
We normally don't guard service calls with local state in polling integrations as we may have the wrong local state, unless there will be a problem calling the service for some state.

MartinHjelmare replied:
Please invert this check and return if true. Then we can outdent below.

nprez83 replied:
What I'm trying to do with this if statement is only send the set_temperature command if the device is not off, otherwise do nothing. That being said, I'm also only returning the target_temperature property if the device is not OFF, which prevents the frontend from allowing a user to even send a set_temperature command if the device is OFF, so we could potentially get rid of the if statement in the set_temperature function altogether.

nprez83 committed:
Update climate.py

MartinHjelmare commented on homeassistant/components/lyric/climate.py:
```diff
@@ -45,7 +47,9 @@
 _LOGGER = logging.getLogger(__name__)
 
-SUPPORT_FLAGS = SUPPORT_TARGET_TEMPERATURE | SUPPORT_PRESET_MODE
+""" Only LCC models support presets """
```
Please use `# ` for code comments.
```suggestion
# Only LCC models support presets
```

MartinHjelmare commented on homeassistant/components/lyric/climate.py:
```diff
@@ -281,23 +291,61 @@ async def async_set_temperature(self, **kwargs) -> None:
     async def async_set_hvac_mode(self, hvac_mode: str) -> None:
         """Set hvac mode."""
-        _LOGGER.debug("Set hvac mode: %s", hvac_mode)
+        _LOGGER.debug("HVAC mode from frontend: %s", hvac_mode)
```
The frontend is not the only thing that can call services.
```suggestion
        _LOGGER.debug("HVAC mode: %s", hvac_mode)
```

nprez83 committed:
Update homeassistant/components/lyric/climate.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

nprez83 committed:
Update homeassistant/components/lyric/climate.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Looks good!

nprez83 commented:
Thanks so much for the help!

MartinHjelmare commented:
Please address the build failures.

nprez83 committed:
Update climate.py

I removed a blank line in 268 and another one at the end of the document. I also fixed the outdents of await commands after the _LOGGER.error calls, not sure what else may be driving the flake8 and black errors. Any guidance is much appreciated @MartinHjelmare

nprez83 committed:
Update climate.py

nprez83 committed:
Update climate.py

corrected some indents that I think were the culprit of the flake8 errors

nprez83 committed:
Update climate.py

I used VS Code to fix locate the flake8 errors. I ran black on it, so I'm hoping that will fix the last lingering black error.
